### PR TITLE
Introduce an `IndexSettingsProvider` to inject logsdb index mode

### LINKED ISSUE #113507: Change the mechanism to enable LogsDB
### Description

With PR https://github.com/elastic/elasticsearch/pull/109025 we introduced a mechanism relying on component templates
to enable LogsDB. This mechanism uses a static setting and despite being useful to enable LogsDB for new deployments
it makes it difficult to do so for existing deployments. Moreover, changing component templates
like `logs@settings` requires bumping the stack template registry version. As a result, enabling LogsDB requires pushing
a change and redeploying. This makes LogsDB activation quite involved.

We decided to introduce a new mechanism that makes enabling LogsDB easier and that does not rely on templates
and/or component templates. We can use the existing `IndexSettingProvider` functionality to evaluate a cluster setting
`cluster.logsdb.enabled` at index creation time and, under certain conditions, inject the `index.mode` setting with value
`logsdb`.

### PR DESCRIPTION
**Pull request by salvatore-campagna**
Here we introduce a new implementation of `IndexSettingProvider` whose goal is to "inject" the
`index.mode` setting with value `logsdb` when a cluster setting `cluster.logsdb.enabled` is `true`.
We also make sure that:
* the existing `index.mode` is not set
* the datastream name matches the `logs-*-*` pattern
* `logs@settings` component template is used

Resolves #113507

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.List;
+
+public class LogsdbIndexModeSettingsProvider implements IndexSettingProvider {
+    private static final Logger logger = LogManager.getLogger(LogsdbIndexModeSettingsProvider.class);
+    private static final PatternMatcher logsNameMatcher = PatternMatcher.forLogs();
+
+    public static final Setting<Boolean> CLUSTER_INDEX_MODE_LOGSDB_ENABLED = Setting.boolSetting(
+        "cluster.index.mode.logsdb.enabled",
```
Comment (salvatore-campagna) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:29]: I wanted to use a different setting to avoid reusing an existing one. I think it would be better to remove the `cluster.logsdb.enabled` setting.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/PatternMatcher.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.datastreams.logsdb;
+
+import java.util.regex.Pattern;
+
+public class PatternMatcher {
```
Comment (martijnvg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/PatternMatcher.java:14]: I don't think this is needed. We already have logic that deals with pattern matches for index templates. I think just using is good enough: `Regex.simpleMatch(logs-*-*, dataStreamName`)`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        final Metadata metadata,
+        final Instant resolvedAt,
+        final Settings settings,
+        final List<CompressedXContent> combinedTemplateMappings
+    ) {
+        if (isLogsdbEnabled == false) {
+            return Settings.EMPTY;
+        }
+
+        final IndexMode indexMode = resolveIndexMode(settings.get("index.mode"));
```
Comment (martijnvg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:61]: I don't think we want to overwrite index mode, even if index.mode=standard is set. I think we should only resolve it when it isn't defined. So maybe just check if index.mode is set and only then based on data stream name decide whether logsdb index mode should be used?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java (hunk 1 above)
Comment (martijnvg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:29]: What is wrong with the existing setting? Maybe just move it to here?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    public static final Setting<Boolean> CLUSTER_INDEX_MODE_LOGSDB_ENABLED = Setting.boolSetting(
+        "cluster.index.mode.logsdb.enabled",
+        false,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    private boolean isLogsdbEnabled;
+
+    public LogsdbIndexModeSettingsProvider(final ClusterService clusterService) {
```
Comment (martijnvg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:36]: Let's remove the `ClusterService` dependency from this class? This would make unit testing this class easier.
We can do `clusterService.getClusterSettings().addSettingsUpdateConsumer(...)` from the plugin class. The setter is already defined here.

**Review by martijnvg — COMMENTED** [review]
I left a few comments. Like we discussed this index settings provider should move to do soon to be added logsdb xpack plugin.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java (hunk 1 above)
Comment (salvatore-campagna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:29]: I wanted to avoid having the same setting controlling two different things.

Comment (martijnvg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:29]: The old mechanism doesn't work and I think we should remove it. Integrations can either hardcode index.mode=logsdb in index templates or they can rely on this index settings provider to enable logsdb.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java (hunk 3 above)
Comment (salvatore-campagna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:61]: I understand that the `IndexSettingsProvier` mechanism kicks in after template composition happens...if we do what you suggest and the template composition comes with `index.mode: standard` than we would not overwrite it with `logsdb`. Is this what we want?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/PatternMatcher.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    private PatternMatcher(final Pattern pattern) {
+        this.pattern = pattern;
+    }
+
+    public boolean matches(final String value) {
+        return value != null && pattern.matcher(value).matches();
+    }
+
+    public static PatternMatcher forLogs() {
+        return new PatternMatcher(Pattern.compile("logs-[^-]+-[^-]+"));
```
Comment (felixbarny) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/PatternMatcher.java:26]: This doesn't match `logs-generic-my-namespace`. Whether dashes are allowed in the namespace is somewhat ambiguous. We should assume that this can happen.

```suggestion
        return new PatternMatcher(Pattern.compile("logs-.+-.+"));
```

Comment (salvatore-campagna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/PatternMatcher.java]: ok...I thought that we required exactly 2 "-"...is this documented somewhere?

Comment (salvatore-campagna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/PatternMatcher.java]: Anyway, the regex you suggested does not work...I tested it versus "logs-apache-production" I will replace it with a working one.

Comment (felixbarny) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/PatternMatcher.java]: The index template doesn't have that as a requirement:

https://github.com/elastic/elasticsearch/blob/a8d7f79070df0fe58ebe16a4197493c6e9210dae/x-pack/plugin/core/template-resources/src/main/resources/logs%40template.json#L2

See also https://github.com/elastic/elasticsearch/issues/112936#issuecomment-2368076975 for a discussion about the discrepancy in the ECS spec: https://github.com/elastic/elasticsearch/issues/112936#issuecomment-2368076975.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java (hunk 3 above)
Comment (martijnvg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:66]: Yes, I think we should only configure `index.mode` to `logsdb` when it hasn't been set already.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/PatternMatcher.java (hunk 5 above)
Comment (martijnvg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/PatternMatcher.java]: This class is not needed and has already been removed from this PR.

 `Regex.simpleMatch(logs-*-*, dataStreamName)`` is now used instead, which does exactly what happens when index templates are resolved.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+public class LogsdbIndexModeSettingsProvider implements IndexSettingProvider {
+    private static final Logger logger = LogManager.getLogger(LogsdbIndexModeSettingsProvider.class);
+
+    public static final Setting<Boolean> CLUSTER_LOGSDB_ENABLED = Setting.boolSetting(
+        "cluster.logsdb.enabled",
+        false,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    public static final String LOGS_PATTERN = "logs-.+-.+";
```
Comment (felixbarny) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:36]: Since this is used in the context of `Regex.simpleMatch`, I think this needs to be adapted
```suggestion
    public static final String LOGS_PATTERN = "logs-*-*";
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+public class LogsdbIndexModeSettingsProvider implements IndexSettingProvider {
+    private static final Logger logger = LogManager.getLogger(LogsdbIndexModeSettingsProvider.class);
+
+    public static final Setting<Boolean> CLUSTER_LOGSDB_ENABLED = Setting.boolSetting(
+        "cluster.logsdb.enabled",
+        false,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    private static final Pattern LOGS_PATTERN = Pattern.compile("logs-.+-.+");
```
Comment (martijnvg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:36]: Why can't we use `Regex.simpleMatch(logs-*-*, dataStreamName)` instead? The `Regex#simpelMatch(...)` is also used to match index templates when creating indices.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java [hunk 8]
```diff
@@ -131,6 +132,8 @@
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
 
+import static org.elasticsearch.datastreams.logsdb.LogsdbIndexModeSettingsProvider.CLUSTER_LOGSDB_ENABLED;
```
[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java:135]: It was here waiting for the other pr to go in dding logsdb xpack plugin.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java (hunk 3 above)
Comment (salvatore-campagna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/datastreams/logsdb/LogsdbIndexModeSettingsProvider.java:66]:
```
final IndexMode indexMode = resolveIndexMode(settings.get("index.mode"));
        if (indexMode != null) {
            return Settings.EMPTY;
        }
```

This is what I do...I check it for `null`

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java [hunk 9]
```diff
@@ -17,13 +19,16 @@
...
 
+import static org.elasticsearch.xpack.cluster.settings.ClusterSettings.CLUSTER_LOGSDB_ENABLED;
 import static org.elasticsearch.xpack.logsdb.SyntheticSourceLicenseService.FALLBACK_SETTING;
 
 public class LogsDBPlugin extends Plugin {
 
     private final Settings settings;
     private final SyntheticSourceLicenseService licenseService;
 
+    private final SetOnce<LogsdbIndexModeSettingsProvider> logsdbIndexModeSettingsProvider = new SetOnce<>();
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:30]: Also we need to return `logsdbIndexModeSettingsProvider` in `getAdditionalIndexSettingProviders()` method?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/settings/ClusterSettings.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.cluster.settings;
+
+import org.elasticsearch.common.settings.Setting;
+
+public class ClusterSettings {
+    public static final Setting<Boolean> CLUSTER_LOGSDB_ENABLED = Setting.boolSetting(
```
Comment (martijnvg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/settings/ClusterSettings.java:13]: Also move this to logsdb plugin? In either to index setting provider or plugin class?

File: x-pack/plugin/core/src/main/java/module-info.java [hunk 11]
```diff
@@ -228,6 +228,7 @@
     exports org.elasticsearch.xpack.core.watcher.trigger;
     exports org.elasticsearch.xpack.core.watcher.watch;
     exports org.elasticsearch.xpack.core.watcher;
+    exports org.elasticsearch.xpack.cluster.settings;
```
[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply x-pack/plugin/core/src/main/java/module-info.java:231]: That package is where I added the `ClusterSettings` class where the shared `CLUSTER_LOGSDB_ENABLED` setting is available. This is required because now that setting is used by `x-pack:stack` and `x-pack:logsdb`...so I need it to be in a common place and the package being exported.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/settings/ClusterSettings.java (hunk 10 above)
Comment (salvatore-campagna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/settings/ClusterSettings.java:13]: See comment about the shared setting.

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+import org.elasticsearch.index.IndexSettingProvider;
+
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.cluster.settings.ClusterSettings.CLUSTER_LOGSDB_ENABLED;
+
+public class LogsdbIndexModeSettingsProvider implements IndexSettingProvider {
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java:28]: I think this class can be final and have package protected visibility?

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java [hunk 13]
```diff
@@ -17,23 +17,31 @@
...
 
+import static org.elasticsearch.xpack.cluster.settings.ClusterSettings.CLUSTER_LOGSDB_ENABLED;
 import static org.elasticsearch.xpack.logsdb.SyntheticSourceLicenseService.FALLBACK_SETTING;
 
 public class LogsDBPlugin extends Plugin {
 
     private final Settings settings;
     private final SyntheticSourceLicenseService licenseService;
 
+    private final LogsdbIndexModeSettingsProvider logsdbIndexModeSettingsProvider;
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:28]: I think this needs to be returned by: `getAdditionalIndexSettingProviders()` method in this class?

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java [hunk 14]
```diff
@@ -17,23 +17,31 @@
...
         licenseService.setLicenseState(XPackPlugin.getSharedLicenseState());
         var clusterSettings = services.clusterService().getClusterSettings();
         clusterSettings.addSettingsUpdateConsumer(FALLBACK_SETTING, licenseService::setSyntheticSourceFallback);
+        clusterSettings.addSettingsUpdateConsumer(
+            CLUSTER_LOGSDB_ENABLED,
+            logsdbIndexModeSettingsProvider::updateClusterIndexModeLogsdbEnabled
+        );
         // Nothing to share here:
         return super.createComponents(services);
     }
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:47]: Also the `CLUSTER_LOGSDB_ENABLED` setting needs to be returned  by the `getSettings(...)` method? (instead of the stack plugin?)

**Review by martijnvg — COMMENTED** [review]
Maybe add a rest level integration test that check that the setting is injected when logsbd.enable setting is enabled?

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java (hunk 9 above)
Comment (salvatore-campagna) [reply x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java]: Yes I need to write a bunch of tests that exercise the full thing other than just calling `getAdditionalIndexSettings`.

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java (hunk 13 above)
Comment (salvatore-campagna) [reply x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:28]: It is right?

```
return List.of(new SyntheticSourceIndexSettingsProvider(licenseService), logsdbIndexModeSettingsProvider);
```

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java (hunk 14 above)
Comment (salvatore-campagna) [reply x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:47]: It looks like that if I include the setting here `CUSTER_LOGSDB_ENABLED` I am not able to run javaRestTest and also the CI results in a lot of failures. I still don't know why. For now I will remove it to submit a few more tests and see if they work.

Comment (salvatore-campagna) [reply x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:47]: So in `SettingsModule` method `registerSetting` we check if we try to register the setting twice.
If we return `CLUSTER_LOGSDB_SETTING` from both `LogsDBPlugin` and `StackPlugin` we end-up registering the same setting twice. I am not sure anyway if deduplicating the list of settings `additionalSettings` in `NodeConstruction` is the right thing to do.

File: server/src/main/java/org/elasticsearch/node/NodeConstruction.java [hunk 15]
```diff
@@ -514,7 +514,7 @@ private SettingsModule validateSettings(Settings envSettings, Settings settings,
         // so we might be late here already
         SettingsModule settingsModule = new SettingsModule(
             settings,
-            additionalSettings,
+            additionalSettings.stream().distinct().collect(Collectors.toList()),
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/node/NodeConstruction.java:517]: Why is this needed? We should not have duplicate settings

**Review by rjernst — COMMENTED** [review]
Is the PR title based on a previous implementation? I do not see anything called an `IndexSettingsProvider` here?

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/LogsIndexModeDisabledRestTestIT.java [hunk 16]
```diff
@@ -54,4 +70,31 @@ public void testLogsSettingsIndexModeDisabled() throws IOException {
         assertThat(indexMode, Matchers.not(equalTo(IndexMode.LOGSDB.getName())));
     }
 
+    public void testTogglingLogsdb() throws IOException {
```
Comment (martijnvg) [inline modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/LogsIndexModeDisabledRestTestIT.java:73]: Maybe move these tests to logsdb module?

File: server/src/main/java/org/elasticsearch/node/NodeConstruction.java (hunk 15 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/node/NodeConstruction.java:517]: Yes, I don't think a change to this file is needed.

File: x-pack/plugin/core/template-resources/src/main/resources/logs@settings-logsdb.json [hunk 17]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (martijnvg) [inline x-pack/plugin/core/template-resources/src/main/resources/logs@settings-logsdb.json:1]: why are these templates moved to logsdb module? I think these templates should stay in the stack module?

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java (hunk 14 above)
Comment (martijnvg) [reply x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:47]: If we register the `CLUSTER_LOGSDB_ENABLED` setting in logsdb plugin then we need to remove the registration in the stack module. A setting can't be registered more than once.

This comment is more than a nice to have. I'm ok with keeping `CLUSTER_LOGSDB_ENABLED` in the stack module. We can move this cluster setting in a follow up when we remove the logic in stack module around enabling logsdb index mode in templates.

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+
+        return Settings.EMPTY;
+    }
+
+    private static boolean matchesLogsPattern(final String name) {
+        return Regex.simpleMatch(LOGS_PATTERN, name);
+    }
+
+    private IndexMode resolveIndexMode(final String mode) {
+        return mode == null
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java:73]: I don't think there is a need here to use streams api. Maybe just do:

```
 IndexMode indexMode = mode != null ? Enum.valueOf(IndexMode.class, mode(Locale.ROOT)) : null;
```

Since we read `mode` variable from settings, it can only have valid values.

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    private boolean usesLogsAtSettingsComponentTemplate(final Metadata metadata, final String name) {
+        final String template = MetadataIndexTemplateService.findV2Template(metadata, name, false);
+        if (template == null) {
+            return false;
+        }
+        final ComposableIndexTemplate composableIndexTemplate = metadata.templatesV2().get(template);
+        if (composableIndexTemplate == null) {
+            return false;
+        }
+        return composableIndexTemplate.composedOf()
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java:87]: Maybe replace this with a for loop?

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+    private static final Logger logger = LogManager.getLogger(LogsdbIndexModeSettingsProvider.class);
+    private static final String LOGS_PATTERN = "logs-*-*";
+    private volatile boolean isLogsdbEnabled;
+
+    LogsdbIndexModeSettingsProvider(final Settings settings) {
+        this.isLogsdbEnabled = CLUSTER_LOGSDB_ENABLED.get(settings);
+    }
+
+    public void updateClusterIndexModeLogsdbEnabled(boolean isLogsdbEnabled) {
+        logger.debug("LogsDB " + (isLogsdbEnabled ? "enabled" : "disabled"));
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java:38]: Maybe remove this debug log statement? The setting infrastructure already emits an info log when cluster settings are updated.

File: server/src/main/java/org/elasticsearch/node/NodeConstruction.java (hunk 15 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/node/NodeConstruction.java:517]: There is a setting `cluster.logsdb.enabedl` which is used by `StackPlugin` and `LogsDBPlugin`, and which is returned by method `getSettings` by both. SO I have two options here:
1. I deduplicate the additional settings as done here.
2. I return the setting only from one of the plugins.

I saw the code in `NodeConstruction` passing these settings to the `SettingsModule` and I see the logic to register settings which does not allow duplicates. Some of my tests were failing before introducing this change because of the exception thrown while registering duplicate settings. I see the logic uses a list of settings rather than a set which tells me that for some reason duplicates are not expected.

For my understanding would be useful to know why duplicates are not expected.

@rjernst what is the right way to proceed? Is there any reason why we can't return the same settings from two different plugins and deduplicate.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/node/NodeConstruction.java:517]: > I return the setting only from one of the plugins.

This is the only option. Registering the same setting twice isn't allowed because this would allow for leniency.

File: x-pack/plugin/core/template-resources/src/main/resources/logs@settings-logsdb.json (hunk 17 above)
Comment (salvatore-campagna) [reply x-pack/plugin/core/template-resources/src/main/resources/logs@settings-logsdb.json:1]: The new `logs@settings-logsdb` template is required because we can have two cases:
1. "mode": "logsdb" when logsdb is enabled
2. no `index.mode` at all. In `LogsdbIndexModeSettingsProvider` the logic expect to have empty/null `index.mode` to inject the "logsdb" index mode.

The way it is the template now always assigns `index.mode` a value, "logsdb" or "standard". As a result injection would never work in `LogsdbIndexModeSettingsProvider`.

I believe if this setting does not work in `StackTemplateRegistry` is it better to remove it...despite requiring more work...especially because some tests might rely on it. If you agree I would do the removal in a separate PR to keep the scope of this one a bit more limited.

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/LogsIndexModeDisabledRestTestIT.java (hunk 16 above)
Comment (martijnvg) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/LogsIndexModeDisabledRestTestIT.java:73]: Apologies  for the confusion. I was just referring to the new tests that were added.

Comment (martijnvg) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/LogsIndexModeDisabledRestTestIT.java:73]: To be more specific I mean moving `testTogglingLogsdb()`, `testEnablingLogsdb()`, `testTogglingLogsdb()` and `testLogsAtSettingWithTimeSeriesOverride()` to `LogsdbRestIT` in the logsdb module.

File: x-pack/plugin/core/template-resources/src/main/resources/logs@settings-logsdb.json (hunk 17 above)
Comment (martijnvg) [reply x-pack/plugin/core/template-resources/src/main/resources/logs@settings-logsdb.json:1]: Ok, so we said we temporarily add this template file so the `xpack.stack.template.logsdb.index.mode` template variable can be removed. Otherwise standard would always be set as index mode and there is no way to know whether this was set specifically or a default.

After this PR is merged we will remove the `cluster.logsdb.enable` setting logic and this templayte from the stack module. Which makes reviewing that change easier to review.

File: x-pack/plugin/stack/src/main/java/org/elasticsearch/xpack/stack/StackPlugin.java [hunk 21]
```diff
@@ -14,6 +14,8 @@
 import java.util.Collection;
 import java.util.List;
 
+import static org.elasticsearch.xpack.cluster.settings.ClusterSettings.CLUSTER_LOGSDB_ENABLED;
```
Comment (martijnvg) [inline x-pack/plugin/stack/src/main/java/org/elasticsearch/xpack/stack/StackPlugin.java:17]: maybe undo this change? It isn't needed and reduce that files this PR touches.

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/LogsIndexModeDisabledRestTestIT.java (hunk 16 above)
Comment (salvatore-campagna) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/LogsIndexModeDisabledRestTestIT.java:73]: Those methods use stuff that is in base class `LogsIndexModeRestTestIT`. So I either move all clasees ot I need to duplicate code creating another base class there or methods used by the test.

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java (hunk 14 above)
Comment (martijnvg) [reply x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:47]: Setting is now only registered in stack module.

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+        final Metadata metadata,
+        final Instant resolvedAt,
+        final Settings settings,
+        final List<CompressedXContent> combinedTemplateMappings
+    ) {
+        if (isLogsdbEnabled == false || dataStreamName == null) {
+            return Settings.EMPTY;
+        }
+
+        final IndexMode indexMode = resolveIndexMode(settings.get("index.mode"));
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java:55]: maybe use `IndexSetting.MODE.getKey()` instead of `index.mode` directly?

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/LogsIndexModeDisabledRestTestIT.java (hunk 16 above)
Comment (martijnvg) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/LogsIndexModeDisabledRestTestIT.java:73]: Ok, in that case let's leave the tests where they are. In a follow up change we can move all logsdb integration tests to the logsdb module.

File: x-pack/plugin/stack/src/main/java/org/elasticsearch/xpack/stack/StackPlugin.java (hunk 21 above)
Comment (salvatore-campagna) [reply x-pack/plugin/stack/src/main/java/org/elasticsearch/xpack/stack/StackPlugin.java:17]: If I undo this I need to register the setting in `LogsDBPlugin`...is that what you are asking?

File: x-pack/plugin/logsdb/src/test/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProviderTests.java [hunk 23]
```diff
@@ -0,0 +1,324 @@
...
+            null,
+            "logs-apache-production",
+            false,
+            Metadata.EMPTY_METADATA,
+            Instant.now().truncatedTo(ChronoUnit.SECONDS),
+            Settings.EMPTY,
+            List.of(new CompressedXContent(DEFAULT_MAPPING))
+        );
+
+        assertIndexMode(additionalIndexSettings, null);
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/test/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProviderTests.java:59]: In this case the returned settings should be empty. Maybe assert whether `additionalIndexSettings.isEmpty()` is true instead?


Maybe all `assertIndexMode(additionalIndexSettings, null);` imstances can be replaced with `assertTrue(additionalIndexSettings.isEmpty())`

File: x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+
+final class LogsdbIndexModeSettingsProvider implements IndexSettingProvider {
+    private static final String LOGS_PATTERN = "logs-*-*";
+    private volatile boolean isLogsdbEnabled;
+
+    LogsdbIndexModeSettingsProvider(final Settings settings) {
+        this.isLogsdbEnabled = CLUSTER_LOGSDB_ENABLED.get(settings);
+    }
+
+    public void updateClusterIndexModeLogsdbEnabled(boolean isLogsdbEnabled) {
```
Comment (martijnvg) [inline x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java:33]: This methods visibility can also be package protected?

**Review by martijnvg — APPROVED** [review]
I think there are three outstanding comments, otherwise LGTM.

**Comment by salvatore-campagna** [comment]
> Is the PR title based on a previous implementation? I do not see anything called an `IndexSettingsProvider` here?

`LogsdbIndexModeSettingsProvider` implements `IndexSettingProvider`...there is no previous implementation.
